week7/cooper_client_react/02_add_login.md

```diff
@@ -0,0 +1,418 @@
+    await storeAuthCredentials(response)
+    return { authenticated: true }
+  } catch (error) {
+    return { authenticated: false, message: error.response.data.errors[0] }
+  }
+};
+
+const storeAuthCredentials = ({ data, headers }) => {
+  return new Promise((resolve) => {
+    const uid = headers['uid'],
+      client = headers['client'],
+      accessToken = headers['access-token'],
+      expiry = headers['expiry'];
+
+    sessionStorage.setItem('credentials', JSON.stringify({
+      uid: uid,
+      client: client,
+      access_token: accessToken,
+      expiry: expiry,
+      token_type: 'Bearer'
+    }));
+    sessionStorage.setItem('current_user', JSON.stringify({ id: data.data.id }));
+    resolve(true)
+  })
+};
+
+export { authenticate, storeAuthHeaders }
+
+```
+
+You will also need to add the axios package with npm, run `npm i axios -S`
+
+We also need to import this to the `App` component:
+
+```js
+import { authenticate } from './Modules/Auth';
+
+```
+
+In the backend, make sure that you have a user that has `johndoe@mail.com` as the email and `password` as the password.
+
+If you run the test and have the backend running locally, you will see that we still get the same error message about how it can't find the text that we are expecting to see. But, we got rid of the error message that we got when we ran the application about how `authenticate` is not defined. If you take a look at the terminal window where you run the backend, you can actually see that we hit it two times, one successful and one not successful. Now it just a question of fetching the response and display it to the user.
+
+So, if we login successfully and store the user in the `sessionStorage`, we want to to display `Hi johndoe@mail.com`. On the other hand, if the login is not successful, we send back the message we get from the response back to the `onLogin` function and store it in a state called message. We want to display that message our page. 
+
+Modify the code in the `render` method for the `App` component to look like this:
+
+```js
+render() {
+    let renderLogin;
+    let user;
+
+    if (this.state.authenticated === true) {
+      user = JSON.parse(sessionStorage.getItem('credentials')).uid;
+      renderLogin = (
+        <p>Hi {user}</p>
+      )
+    } else {
+      if (this.state.renderLoginForm === true) {
+        renderLogin = (
+          <>
+            <LoginForm 
+              loginHandler={this.onLogin.bind(this)}
+              inputChangeHandler={this.onChange.bind(this)}
+            />
+          </>
+        )
+      } else {
+        renderLogin = (
+          <>
+            <button id="login" onClick={() => this.setState({ renderLoginForm: true })}>Login</button>
+            <p>{this.state.message}</p>
+          </>
+        )
+      }
+    }
+    
+    return (
+      // <...>
+    );
+  }
+
+```
+If you run the tests now with the backend running, everything should go green!
```

Comment:
If not make sure that you go through the code again, and start up the application as well....
kanske

Reply:
done